src/redux/note-details/reducer.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { Reducer } from 'redux'
+import { PresentFrontmatterExtractionResult } from '../../components/common/note-frontmatter/types'
+import { NoteFrontmatter } from '../../components/common/note-frontmatter/note-frontmatter'
+import { NoteDetails, NoteDetailsActions, NoteDetailsActionType } from './types'
+import { noteDtoToNoteDetails } from '../../api/notes/dto-methods'
+import { extractFrontmatter } from '../../components/common/note-frontmatter/extract-frontmatter'
+import { NoteDto } from '../../api/notes/types'
+import { initialState } from './initial-state'
+
+export const NoteDetailsReducer: Reducer<NoteDetails, NoteDetailsActions> = (
+  state: NoteDetails = initialState,
+  action: NoteDetailsActions
+) => {
+  switch (action.type) {
+    case NoteDetailsActionType.SET_DOCUMENT_CONTENT:
+      return buildStateFromDocumentContentUpdate(state, action.content)
+    case NoteDetailsActionType.UPDATE_NOTE_TITLE_BY_FIRST_HEADING:
+      return buildStateFromFirstHeadingUpdate(state, action.firstHeading)
+    case NoteDetailsActionType.SET_NOTE_DATA_FROM_SERVER:
+      return buildStateFromServerDto(state, action.dto)
+    case NoteDetailsActionType.UPDATE_TASK_LIST_CHECKBOX:
+      return buildStateFromTaskListUpdate(state, action.changedLine, action.checkboxChecked)
+    default:
+      return state
+  }
+}
+
+const TASK_REGEX = /(\s*(?:[-*+]|\d+[.)]) )(\[[ xX]])( .*)/
+
+/**
+ * Builds a {@link NoteDetails} redux state from a DTO received as an API response.
+ * @param state The previous redux state.
+ * @param dto The first DTO received from the API containing the relevant information about the note.
+ * @return An updated {@link NoteDetails} redux state.
+ */
+const buildStateFromServerDto = (state: NoteDetails, dto: NoteDto): NoteDetails => {
+  const newState = {
+    ...state,
```

Comment:
Why are you preserving the old state here? In fact the convert method overwrites every attribute.